src/js/services/optimizely.test.js

```diff
@@ -0,0 +1,265 @@
+  afterEach(() => {
+    jest.clearAllMocks();
+    jest.resetAllMocks();
+  });
+
+  describe('getUserId()', () => {
+    beforeEach(() => {
+      // Ensure glob is cleaned up after each test
+      glob.userData = null;
+    });
+
+    it('test if userData exists but analytics_id undefined', () => {
+      glob.userData = {};
+      return client.getUserId().then((data) => {
+        expect(data).toBe(null);
+      });
+    });
+
+    it('test if userData exists and analytics_id not null', async () => {
+      glob.userData = {
+        analytics_id: '00000000-0000-0000-0000-000000000000',
+      };
+      await expect(client.getUserId()).resolves.toBe(
+        '00000000-0000-0000-0000-000000000000',
+      );
+    });
+
+    it('test waiting for userData to populate null', async () => {
+      glob.userData = null;
+
+      setTimeout(() => {
+        glob.userData = {
+          analytics_id: undefined,
+        };
+        window.dispatchEvent(userDataReady);
+      }, 100);
+
+      await expect(client.getUserId()).resolves.toBe(null);
+    });
+
+    it('test if userData does not exist initially and defined later', async () => {
+      glob.userData = null;
+      setTimeout(() => {
+        glob.userData = {
+          analytics_id: '00000000-0000-0000-0000-000000000000',
+        };
+        window.dispatchEvent(userDataReady);
+      }, 100);
+
+      await expect(client.getUserId()).resolves.toBe(
+        '00000000-0000-0000-0000-000000000000',
+      );
+    });
+  });
+
+  // Function used in TrackExperimentViewed
+  describe('isExperimentAlreadyViewed()', () => {
+    beforeEach(() => {
+      // ensure local storage clear
+      window.localStorage.clear();
+    });
+    it('experiment has not been viewed before', () => {
+      storeExperimentParticipation(orgId, experimentKey + '456', variationName);
+      expect(isExperimentAlreadyViewed(orgId, experimentKey)).toBe(false);
+    });
+    it('experiment has been viewed before', () => {
+      storeExperimentParticipation(orgId, experimentKey, variationName);
+      expect(isExperimentAlreadyViewed(orgId, experimentKey)).toBe(true);
+    });
+  });
+
+  //Function used in getVariationName
+  describe('trackExperimentViewed()', () => {
+    beforeEach(() => {
+      global.AnalyticsClient = AnalyticsClient;
+      window.localStorage.clear();
+      jest.clearAllMocks();
+    });
+
+    afterEach(() => {
+      delete global.AnalyticsClient;
+      jest.resetAllMocks();
+    });
+
+    it('expect trackExperimentViewed to have bene viewed', () => {
+      storeExperimentParticipation(orgId, experimentKey, variationName);
+      expect(isExperimentAlreadyViewed(orgId, experimentKey)).toBe(true);
+
+      const spy = jest.spyOn(window.AnalyticsClient, 'trackAction');
+      trackExperimentViewed(
+        orgId,
+        experimentKey,
+        ['container item'],
+        '5',
+        variationName,
+        '5',
+        '5',
+      );
+      expect(spy).not.toHaveBeenCalled();
```

Comment:
suggestion: Here we should test that the `trackAction` has been called with the correct parameters

Reply:
It would actually be the second one because the first one skips over trackAction but its a good point!